rust/src/quic/quic.rs

```diff
@@ -112,44 +129,146 @@ impl QuicState {
+            let mut pktnum_buf = Vec::with_capacity(4);
+            pktnum_buf.extend_from_slice(&h2[hlen..hlen + 4]);
+            let r1 = hkey.decrypt_in_place(
+                &h2[hlen + 4..hlen + 4 + hkey.sample_len()],
+                &mut h20,
+                &mut pktnum_buf,
+            );
+            if !r1.is_ok() {
+                return false;
+            }
+            // mutate one at a time
+            h2[0] = h20;
+            let _ = &h2[hlen..hlen + 1 + ((h20 & 3) as usize)]
+                .copy_from_slice(&pktnum_buf[..1 + ((h20 & 3) as usize)]);
+            let pkt_num = quic_pkt_num(&h2[hlen..hlen + 1 + ((h20 & 3) as usize)]);
+            if framebuf.len() < 1 + ((h20 & 3) as usize) {
+                return false;
+            }
+            output.extend_from_slice(&framebuf[1 + ((h20 & 3) as usize)..]);
+            let pkey = if to_server {
+                &keys.remote.packet
+            } else {
+                &keys.local.packet
+            };
+            let r = pkey.decrypt_in_place(pkt_num, &h2[..hlen + 1 + ((h20 & 3) as usize)], output);
+            if !r.is_ok() {
+                return false;
+            }
+            return true;
+        }
+        // unreachable by the way
+        return false;
+    }
+
+    fn parse(&mut self, input: &[u8], to_server: bool) -> bool {
+        // so as to loop over multiple quic headers in one packet
+        let mut buf = input;
+        while buf.len() > 0 {
+            match QuicHeader::from_bytes(buf, DEFAULT_DCID_LEN) {
+                Ok((rest, header)) => {
+                    // unprotect
+                    if self.hello_tc && self.hello_ts {
+                        // payload is encrypted, stop parsing here
+                        return true;
+                    }
+                    if self.keys.is_none() && header.ty == QuicType::Initial {
+                        if let Some(version) = quic_rustls_version(u32::from(header.version)) {
+                            let keys = rustls::quic::Keys::initial(version, &header.dcid, false);
+                            self.keys = Some(keys)
+                        }
+                    }
+                    if header.length > rest.len() {
+                        //TODO event whenever an error condition is met
+                        return false;
+                    }
+                    let (mut framebuf, next_buf) = rest.split_at(header.length);
+                    let hlen = buf.len() - rest.len();
+                    let mut output;
+                    if self.keys.is_some() {
+                        output = Vec::with_capacity(framebuf.len() + 4);
+                        if !self.decrypt(to_server, &header, framebuf, buf, hlen, &mut output) {
+                            return false;
+                        }
+                        framebuf = &output;
+                    }
+                    buf = next_buf;
+                    match QuicData::from_bytes(framebuf) {
```

Comment:
can this func be broken up, for example handling the following in a until func? Not only is the func large, the indent levels are also getting hard to track

Reply:
ok